### ISSUE TITLE
Selection of best checkpoint without using data labels?

### ISSUE THREAD

**Issue by ssgosh** [inspire-group/SSD#7]
I am trying to train this Outlier Detection model using a custom, unsupervised dataset, using the "SimCLR" training mode. However, I noticed that both validation methods used in the code (knn and baseeval) require data labels. 

Is it possible to select the "best" checkpoint from the training dataset without using labelled data?

**Comment by VSehwag** [comment]
This is one of most fundamental questions in unsupervised learning: "How to test the quality of learned embeddings?" For eg., let's consider clustering. How do we test if the clustering output is good, if we don't have the ground truth labels. In such cases folks resort to heuristic approaches, e.g., are clusters well separated and dense. So there is no easy solution when there are no ground truth labels and one has to use some domain knowledge to develop a custom proxy test.

**Comment by ssgosh** [comment]
How about selecting the checkpoint for which the average anomaly score of the validation data is minimized (assuming there are no anomalies in the validation data)? This would not require any labels. This is the criterion used at test time to determine if the data is typical or an anomaly, so it makes sense that it may be used at validation time.

For the example of clustering, one can assign each validation data point to a training cluster, and verify if the means of each validation cluster is close to the mean of each training cluster.

**Comment by VSehwag** [comment]
Ofc. One can design such validation set to determine whether the learned embedding are good for anomaly detection (note that labels are still there but they are just binary labels, i.e., inlier/outlier). Make sure to use a separate validation data for such analyses.
